File: samples/android/src/main/java/org/gosuda/portal/sample/AdvancedContract.kt
package org.gosuda.portal.sample

import android.content.Context
import org.gosuda.portal.Capability
import org.gosuda.portal.PortalClient
import org.gosuda.portal.PortalConfig
import org.gosuda.portal.PortalTunnel

/**
 * Compile-only contract for the advanced low-level path.
 *
 * The sample's primary flow uses the intent factories and `publish`; this
 * fixture keeps the raw `PortalConfig` constructor, `open` (accepted-before-
 * ready), `awaitReady`, and the authoritative `state` stream compiled so the
 * escape hatch cannot drift from the exported API.
 */
internal object AdvancedContract {

    /**
     * Opens a tunnel without waiting for readiness: the caller observes
     * CONNECTING through `state` and awaits a single capability itself.
     */
    suspend fun openAndObserve(context: Context): Pair<PortalClient, PortalTunnel> {
        val client = PortalClient(context.applicationContext)
        val tunnel = client.open(
            PortalConfig(
                name = "advanced-site",
                staticDir = "/data/site",
                staticIndex = "index.html",
                discovery = true
            )
        )
        tunnel.state.collect { snapshot ->
            // Render CONNECTING/ACTIVE transitions and relay updates live.
            println("${snapshot.phase} ${snapshot.primaryPublicUrl}")
        }
        tunnel.awaitReady(Capability.STATIC_SITE, timeoutMillis = 15_000)
        return client to tunnel
    }
}

File: samples/ios/AdvancedContract.swift
import Foundation
import PortalSDK

/// Compile-only contract for the advanced low-level path.
///
/// The sample's primary flow uses `PortalIosConfigFactory` and `publish`;
/// this fixture keeps the raw `PortalConfig` initializer, `open`
/// (accepted-before-ready), `awaitReady`, and `observeState` compiled so the
/// escape hatch cannot drift from the exported API.
final class AdvancedContract {

    private let client = PortalIosClient(allowRemoteTargets: false, defaultIdentityPath: nil)
    private var operation: PortalOperation?
    private var subscription: PortalSubscription?

    /// Opens a tunnel without waiting for readiness: the caller observes
    /// CONNECTING through `observeState` and awaits a capability itself.
    func openAndObserve() {
        let config = PortalConfig(
            name: "advanced-site",
            identityJson: nil,
            identityPath: nil,
            relays: nil,
            discovery: KotlinBoolean(bool: true),
            maxActiveRelays: 3,
            banMitm: false, ech: false, overlay: false,
            udp: false, tcp: false,
            description: nil, tags: nil, owner: nil, thumbnail: nil, hide: false,
            staticDir: "/data/site", staticIndex: "index.html",
            targetAddr: nil, udpAddr: nil, httpRoutes: nil, x402: nil
        )
        operation = client.open(config: config) { [weak self] session, failure in
            guard let self else { return }
            if let failure {
                NSLog("portal open failed: %@ %@", failure.code, failure.message)
                return
            }
            subscription = session?.observeState { snapshot in
                NSLog("portal %@ %@", "\(snapshot.phase)", snapshot.primaryPublicUrl ?? "")
            }
            session?.awaitReady(capability: .staticSite, timeoutMillis: 15_000) { _, failure in
                if let failure { NSLog("awaitReady failed: %@", failure.code) }
            }
        }
    }
}
